py/_io/terminalwriter.py

```diff
@@ -144,12 +144,30 @@ def __init__(self, file=None, stringio=False, encoding=None):
     def fullwidth(self):
         if hasattr(self, '_terminal_width'):
             return self._terminal_width
-        return get_terminal_width()
+        try:
+            width = self._cached_terminal_width
+        except AttributeError:
+            try:
+                signal.signal(signal.SIGWINCH, self._handle_sigwinch)
+            except AttributeError:  # SIGWINCH not supported.
+                width = self._cached_terminal_width = False
```

Comment:
> after taking another looks i propose creating a global "cache" as per object basis will overwrite the signals of other terminal-writer objects and cause multiple objects not to get invalidated

Not quite sure I get your proposal, but why would it need to be a global object "cache"? I would prefer to keep the "cache instance" in this instance it self, no need to mess with globals.

But do we need to handle more signals types in other places? If so the idea of separating the "cache value until a signal invalidates it" logic into a separate class and reuse it sounds good, otherwise it might be overkill if it is used only by the `TerminalWriter` and only for `SIGWINCH`.

> in addition it may break interaction with other libraries as is since we don't call existing handlers

Oh I was under the impression that `signal.signal` would *add* a callback, not overwrite it (like `atexit` does add callbacks). I agree we should account for that detail.